### ISSUE TITLE
Add option to add duplicate arguments to output params. (using a list)

### ISSUE THREAD

**Issue by Thomasedv** [abhiTronix/vidgear#141]
The issue is, we can pass a dict with output arguments, but due to the nature of dictionaries, we are unable to take advantage of commands that can be duplicated, such as `-map`.

## Detailed Description
Writing video, i discovered it was possible to add audio as a secondary input and using map to get video from the python process, and audio from a video file. However, due to the way Writegear output params is implemented, we as users can not utilized this ability. 

### Context
The implementation is up to you, but here goes some ideas from me:

You can make duplicate keys in dicts with a bit of class trickery, however, due to using strip() and str() calls internally and duplicate keys are overwritten there. However tweaking this seems a bit complex from just looking at it. So I suggest to simply allow for a list of arguments to be passed, and added to the cmd list after output parameters inside the method `__startFFmpeg_Process` inside of `writegear.py`. The user may then include their own list of arguments that can have duplicates and potentially any argument that does not have a value. (I don't know if any are for output arguments really, but params like -y and -hide_banner exist.) 
Say we add a new key `list_params` (you can pick a better name than `list_param`s though) to the `__init__` of WriterGear:

```
def __init__(
        output_filename="",
        compression_mode=True,
        custom_ffmpeg="",
        logging=False,
        list_params=None
        **output_params
    ):
....
def __startFFmpeg_Process(self, input_params, list_args, output_params):
    ...
    cmd = (
            [self.__ffmpeg, "-y"]
            + ["-f", "rawvideo", "-vcodec", "rawvideo"]
            + input_parameters
            + ["-i", "-"]
            + output_parameters
            + list_params
            + [self.__out_file]
        )

```
This would not break old code, given that if they used positional arguments they would not accidentally set list_params to anything as long as they use the dict unpacking for output_params, as well as all those keys start with a dash since they are for ffmpeg or the specific commands implemented for user convenience like `-input_framerate`. Allowing for this would grant a larger level of freedom when using the writer, without having to resort to a method like in this issue #15 where custom calls have to be made after creating a temp output without sound. 

Edit: From testing, if you provide an input -i in the output params dict, the mapping can be done afterwards in the list_params. See following example code showing the content of both a potential dict and list or params for ffmepg:

```
list_params = ['-map', '0:v:0', '-map', '1:a?', '-acodec', 'libopus', '-b:a', '320k',]
output_params = output_params = {"-input_framerate": str(framerate), '-i': source, '-vcodec': 'libvpx-vp9', '-crf': '25', 'b:v': '40M',}
```

### Your Environment
* VidGear version: 0.1.8
* Branch: PyPi 
* Python version: 3.6
* Operating System and version: Windows 10

**Comment by abhiTronix** [comment]
@Thomasedv Thanks for this idea and explaining it detail. But I think the easiest solution will be to introduce a new attribute for  `output_params` dictionary parameter, that excepts only list as its value. So let's say that attribute we intended _(lets call it `duplicates`)_ can be passed as follows:

```python
output_params =  {'-input_framerate': str(framerate), '-vcodec': 'libvpx-vp9', '-crf': '25', 'b:v': '40M', '-acodec': 'libopus', '-b:a': '320k', 'duplicates': ['-map', '0:v:0', '-map', '1:a?']}
```

and these duplicates will be merged with final list as it is, and will be executed as intended. Therefore serving the required goal here.

**I'll push commit to testing branch soon, please stay tuned.**

**Comment by Thomasedv** [comment]
Sounds good! Will the duplicate list be placed at the spot it's inserted into the ouput_params in that case?

**Comment by abhiTronix** [comment]
> Will the duplicate list be placed at the spot it's inserted into the ouput_params in that case?

Yup.

**Comment by abhiTronix** [comment]
> Edit: From testing, if you provide an input -i in the output params dict, the mapping can be done afterwards in the list_params. See following example code showing the content of both a potential dict and list or params for ffmepg:
> 
> list_params = ['-map', '0:v:0', '-map', '1:a?', '-acodec', 'libopus', '-b:a', '320k',]
> output_params = output_params = {"-input_framerate": str(framerate), '-i': source, '-vcodec': 'libvpx-vp9', '-crf': '25', 'b:v': '40M',}

**@Thomasedv DO NOT give video source at `-i` parameter in `output_params`, otherwise it will interfere with frame input from WriteGear and it will break things.**

**Comment by abhiTronix** [comment]
@Thomasedv  Install `development` branch as follows:

```sh
  # clone the repository and get inside
  git clone https://github.com/abhiTronix/vidgear.git && cd vidgear

  # checkout the latest testing branch
  git checkout developement

  # install normally
  pip install .

  # OR install with asyncio support
  pip install .[asyncio]
```

**and check if it works as you expected or not?**

**Comment by Thomasedv** [comment]
@abhiTronix Getting wrong type error in helper, List is converted to a string. 

Most likely caused by this line not making exception for the list:
https://github.com/abhiTronix/vidgear/blob/ee11ccdd62135a0e50e36ec72cea388cc428d45b/vidgear/gears/writegear.py#L133

**Comment by abhiTronix** [comment]
> Most likely caused by this line not making exception for the list:
> vidgear/vidgear/gears/writegear.py
> 
> Line 133 in ee11ccd
> 
>  str(k).strip().lower(): str(v).strip() for k, v in output_params.items() 
> Update:
> I manually made an exception for -clones to not convert to it's value to string, and can confirm that worked for my use.

@Thomasedv Thanks, updated.

**Comment by Thomasedv** [comment]
I'll do it as soon as I can.

**Comment by Thomasedv** [comment]
@abhiTronix It's working now! 

For reference, i tested with these `ouput_params` (excluding some specific stuff for the video codec), in case anyone else is intending to add audio/subtitles from a video file along with the piped frames. The `source` is the path to a video with audio/subtitles. 

```
output_params = {"-input_framerate": str(framerate),
                         '-i': source,
                         '-clones': ['-map', '0:v:0', '-map', '1:a?', '-map', '1:s?'],
                         '-acodec': 'libopus',
                         '-b:a': '320k',
                         '-vcodec': 'libvpx-vp9',
```

Thanks for the quick implementation!

**Comment by abhiTronix** [comment]
@Thomasedv Thanks a lot for testing. I'll merge this PR in few hours.

**Comment by abhiTronix** [comment]
**Successfully resolved and merged in 52d7659**

**Comment by GuillaumeTong** [comment]
This feature could also be useful in these cases:

- When argument order matter (since python dict do not have guaranteed order) for e.g. in my case, when I want to insert an audio, i will need to specify '-i' before '-vcodec'
- When a flag does not expect any value

This page currently seem to indicate that this feature has been merged into 1.9. I am currently using vidgear 2.1, but I cannot find this feature in the code or documentation. Am I missing something?

**Comment by abhiTronix** [comment]
When argument order matter (since python dict do not have guaranteed order) for e.g. in my case, when I want to insert an audio, i will need to specify '-i' before '-vcodec'
When a flag does not expect any value
This page currently seem to indicate that this feature has been merged into 1.9. I am currently using vidgear 2.1, but I cannot find this feature in the code or documentation. Am I missing something?

@GuillaumeTong Yes, you're missing [`'-clones'`](https://abhitronix.github.io/vidgear/latest/gears/writegear/compression/params/#supported-parameters) attribute of [`output_params`](https://abhitronix.github.io/vidgear/latest/gears/writegear/compression/params/#output_params) dictionary attribute of WriteGear, and it was added way back in `v0.1.9` because of this issue. With `-clones` you can pass parameter as list and also order is not altered when executed. So for example if you do:

```python
output_params = {"-input_framerate": str(framerate),
                         '-i': source,
                         '-clones': ['-map', '0:v:0', '-map', '1:a?', '-map', '1:s?', '-movflags', '+faststart'],
                         '-acodec': 'libfdk_aac',
                         '-b:a': '320k',
                         '-vcodec': 'libvpx-vp9'}

```
Will preserve both duplicates and single(without value) flags as intended.
